### PR TITLE
Dev/rdmmpa api update

### PR DESCRIPTION
**Pull request by bp-kelley**
This PR adds some features to MMPA that allow it to better search MMP's.

Developers can now fragment on a given bond (not just all) and specify the desired number of cuts.  I.e. just give me two cut fragments.  This enables usage in a web-based UI for instance.

```
bool fragmentMol(const ROMol& mol,
                 std::vector<std::pair<ROMOL_SPTR, ROMOL_SPTR> >& result,
                 const std::vector<unsigned int>& bondsToCut,
                 unsigned int minCuts = 1,
                 unsigned int maxCuts = 3);
```

### CODE REVIEW

File: Code/GraphMol/ChemReactions/Wrap/rdChemReactions.cpp [hunk 1]
```diff
@@ -47,6 +47,7 @@
 #include <GraphMol/ChemReactions/ReactionFingerprints.h>
 #include <GraphMol/ChemReactions/ReactionUtils.h>
 #include <boost/python/suite/indexing/vector_indexing_suite.hpp>
+#include <boost/cstdint.hpp>
```
Comment (greglandrum) [inline Code/GraphMol/ChemReactions/Wrap/rdChemReactions.cpp]: ??? Is this relevant to this PR?

File: Code/GraphMol/MMPA/MMPA.cpp [hunk 2]
```diff
@@ -394,8 +408,33 @@ bool fragmentMol(const ROMol& mol,
...
+bool fragmentMol(const ROMol& mol,
+                 std::vector<std::pair<ROMOL_SPTR, ROMOL_SPTR> >& res,
+                 const std::vector<unsigned int>& bondsToCut,
+                 unsigned int minCuts,
+                 unsigned int maxCuts) {
+  std::vector<BondVector_t> matching_bonds;  // List of matched query's bonds
+
+  BOOST_FOREACH(unsigned int i, bondsToCut) {
+    const Bond *bond = mol.getBondWithIdx(i);
+    if(bond) {
```
Comment (greglandrum) [inline Code/GraphMol/MMPA/MMPA.cpp]: This test is unnecessary: `getBondWithIdx()` will either return a result or generate an exception (there's a POSTCONDITION that the result is non-null).

File: Code/GraphMol/MMPA/MMPA.cpp [hunk 3]
```diff
@@ -394,8 +408,33 @@ bool fragmentMol(const ROMol& mol,
...
+      BondVector_t bonds;
+      unsigned int a1 = bond->getBeginAtomIdx();
+      unsigned int a2 = bond->getEndAtomIdx();
+      bonds.push_back( std::pair<unsigned int, unsigned int>(a1, a2) );
```
Comment (greglandrum) [inline Code/GraphMol/MMPA/MMPA.cpp]: `std::make_pair()` is more compact.

File: Code/GraphMol/MMPA/Wrap/rdMMPA.cpp [hunk 4]
```diff
@@ -47,6 +48,116 @@ python::tuple fragmentMolHelper(const RDKit::ROMol& mol, unsigned int maxCuts,
...
+        }
+        lres.append(RDKit::MolToSmiles(*(pr->second), true));
+      }
+      pyres.append(python::tuple(lres));
+    }
+  }
+  return python::tuple(pyres);
+}
+
+template<class T>
```
Comment (greglandrum) [inline Code/GraphMol/MMPA/Wrap/rdMMPA.cpp]: Seems like this does the same thing as `pythonObjectToVect()` in `RDBoost/Wrap.h`

File: Code/GraphMol/MMPA/MMPA.h [hunk 5]
```diff
@@ -36,5 +36,48 @@ bool fragmentMol(const ROMol& mol,
                  std::vector<std::pair<ROMOL_SPTR, ROMOL_SPTR> >& result,
                  unsigned int maxCuts = 3, unsigned int maxCutBonds = 20,
                  const std::string& pattern = "[#6+0;!$(*=,#[!#6])]!@!=!#[*]");
+
+//! fragments a Molecule for processing with the Matched Molecular Pairs
```
Comment (greglandrum) [inline Code/GraphMol/MMPA/MMPA.h]: yay! documentation! thanks!

File: Code/GraphMol/MMPA/Wrap/rdMMPA.cpp [hunk 6]
```diff
@@ -47,6 +48,116 @@ python::tuple fragmentMolHelper(const RDKit::ROMol& mol, unsigned int maxCuts,
...
+  std::vector<unsigned int> vect;
+  size_t num = python::extract<unsigned int>(bbs.attr("__len__")());
+  vect.reserve(num);
+  for(size_t i=0; i<num; ++i) {
+    vect.push_back(python::extract<unsigned int>(bbs[i]));
+  }
+  return vect;
+}
+
+python::tuple fragmentMolHelper3(const RDKit::ROMol& mol,
```
Comment (greglandrum) [inline Code/GraphMol/MMPA/Wrap/rdMMPA.cpp]: I think you can combine this and `fragmentMolHelper4()` by changing the type of the `ob` argument to `python::object` and using `pythonObjectToVect()`, that would allow anything that supports `__len__` and `__getitem__` to be used.

File: Code/GraphMol/MMPA/Wrap/testMMPA.py [hunk 7]
```diff
@@ -1,5 +1,6 @@
+from __future__ import print_function
 from rdkit import RDConfig
-import unittest
+import sys, unittest
```
Comment (greglandrum) [inline Code/GraphMol/MMPA/Wrap/testMMPA.py]: don't need sys here.

File: Code/GraphMol/MMPA/Wrap/testMMPA.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
+from __future__ import print_function
```
Comment (greglandrum) [inline Code/GraphMol/MMPA/Wrap/testMMPA.py]: it's ok to have this here since it may help in the future, but it's not needed for this file at the moment.

File: Code/GraphMol/MMPA/Wrap/testMMPA.py [hunk 9]
```diff
@@ -100,6 +101,53 @@ def test5(self):
...
     self.assertEqual(len(frags), 231)
 
+  def test6(self):
+    m = Chem.MolFromSmiles(
+      "CC[C@H](C)[C@@H](C(=O)N[C@H]1CSSC[C@H]2C(=O)NCC(=O)N3CCC[C@H]3C(=O)N[C@H](C(=O)N[C@H](C(=O)N[C@H](C(=O)N[C@@H](CSSC[C@@H](C(=O)N[C@H](C(=O)N4CCC[C@H]4C(=O)N[C@H](C(=O)N2)C)CC(=O)N)NC1=O)C(=O)N)CO)Cc5ccc(cc5)O)CCCC[NH3+])N")  # ALPHA-CONOTOXIN SI
+    frags = rdMMPA.FragmentMol(m, resultsAsMols=False)
+    self.assertFalse(len(frags))
+    frags1 = rdMMPA.FragmentMol(m, minCuts=1, maxCuts=1, maxCutBonds=21,
+                                resultsAsMols=False)
+    frags2 = rdMMPA.FragmentMol(m, minCuts=2, maxCuts=2, maxCutBonds=21,
```
Comment (greglandrum) [inline Code/GraphMol/MMPA/Wrap/testMMPA.py]: seems like you should also have a test where minCuts and maxCuts are different.

File: Code/GraphMol/MMPA/MMPA.cpp [hunk 10]
```diff
@@ -329,7 +333,17 @@ static inline void processCuts(
...
+                 const std::string& pattern) {
+  return fragmentMol(mol, res, 1, maxCuts, maxCutBonds, pattern);
+}
+
+bool fragmentMol(const ROMol& mol,
+                 std::vector<std::pair<ROMOL_SPTR, ROMOL_SPTR> >& res,
+                 unsigned int minCuts,
+                 unsigned int maxCuts,
+                 unsigned int maxCutBonds,
                  const std::string& pattern) {
```
Comment (greglandrum) [inline Code/GraphMol/MMPA/MMPA.cpp]: seems like you should either have a `PRECONDITION(minCuts<=maxCuts,"min not <= max") on this function or a check that throws a ValueError if that condition isn't true. Maybe the second is better.
Either way including a test for it isn't a bad idea.

**Comment by greglandrum** [comment]
Overall I think this is a useful addition and the API is good.

File: Code/GraphMol/ChemReactions/Wrap/rdChemReactions.cpp (hunk 1 above)
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/Wrap/rdChemReactions.cpp]: Nope.  I'll remove that file.

File: Code/GraphMol/MMPA/Wrap/rdMMPA.cpp (hunk 4 above)
Comment (bp-kelley) [inline Code/GraphMol/MMPA/Wrap/rdMMPA.cpp]: Neat, I missed that.

File: Code/GraphMol/MMPA/Wrap/testMMPA.py (hunk 9 above)
Comment (bp-kelley) [inline Code/GraphMol/MMPA/Wrap/testMMPA.py]: @greglandrum - on it again

Comment (bp-kelley) [inline Code/GraphMol/MMPA/Wrap/testMMPA.py]: The next line does this.  The default minCuts==1 but I can make it more explicit.

Comment (bp-kelley) [inline Code/GraphMol/MMPA/Wrap/testMMPA.py]: basically the checks set(minCuts=maxCuts=1) + set(minCuts+maxCuts=2) == set(mincuts=1,maxCuts=2)

unless I misunderstand your comment.

Comment (greglandrum) [inline Code/GraphMol/MMPA/Wrap/testMMPA.py]: nope, you're right. I missed that since it wasn't explicit. It's fine.

**Comment by bp-kelley** [comment]
@greglandrum Think I touched all your points.

There is a troubling spot in a test though, I've marked it.

**Comment by bp-kelley** [comment]
Apparently I can't.  This line in the test:

 # FIX: this needs to be investigated, it's not currently passing
    #self.assertTrue('O=C(N[*:3])C([*:1])[*:2]' in cores)

Looks suspicious!
